docs/notebooks/rembo.pct.py

```diff
@@ -0,0 +1,246 @@
+def high_dim_objective(x):
+    tf.debugging.assert_shapes([(x, (..., D))])
+
+    return objective(x[..., :2])
+
+
+def build_model(data, d):
+    # add a bit of noise, since there's a risk the variance could be zero for Michalewicz
+    variance = tf.math.reduce_variance(data.observations) + 1e-4
+    kernel = gpflow.kernels.Matern52(variance=variance, lengthscales=[0.2] * d)
+    prior_scale = tf.cast(1.0, dtype=tf.float64)
+    kernel.variance.prior = tfp.distributions.LogNormal(
+        tf.cast(-2.0, dtype=tf.float64), prior_scale
+    )
+    kernel.lengthscales.prior = tfp.distributions.LogNormal(
+        tf.math.log(kernel.lengthscales), prior_scale
+    )
+    gpr = gpflow.models.GPR(data.astuple(), kernel, noise_variance=1e-5)
+    gpflow.set_trainable(gpr.likelihood, False)
+
+    return GaussianProcessRegression(gpr, num_kernel_samples=100)
+
+
+# %% [markdown]
+# ## Run standard Bayesian optimization
+# We run the process 5 times - note that this takes a while!
+
+# %%
+import trieste
+
+final_datasets = []  # to store the results
+
+observer = trieste.objectives.utils.mk_observer(high_dim_objective)
+
+for _ in range(num_seeds):
+    # Sample initial points
+    initial_query_points = search_space.sample_sobol(num_initial_points)
+    initial_data = observer(initial_query_points)
+
+    # Build the model over the high-dimensional space
+    model = build_model(initial_data, d=D)
+
+    # Set up the optimizer and run the loop
+    bo = trieste.bayesian_optimizer.BayesianOptimizer(observer, search_space)
+
+    # Store the results
+    result = bo.optimize(num_steps, initial_data, model)
+    dataset = result.try_get_final_dataset()
+
+    final_datasets.append(dataset)
+
+
+# %% [markdown]
+# We now show how to implement REMBO, by providing a new observer that acts on a projection of the input data by wrapping the original objective.
+
+# %%
+
+
+def make_REMBO_observer_and_search_space(D, d, objective, search_space):
+    assert isinstance(search_space, Box)
+
+    A = tf.random.normal(
+        [D, d], dtype=gpflow.default_float()
+    )  # sample projection matrix
+
+    new_search_space = Box(
+        [-math.sqrt(d)] * d, [math.sqrt(d)] * d
+    )  # recommendation from REMBO paper
+
+    def new_objective(y):
+        tf.debugging.assert_shapes([(y, (..., d))])
+
+        rescaled_search_space = Box(
+            [-1.0] * D, [1.0] * D
+        )  # REMBO assumes the original space has bounds [-1, 1]^D
+
+        scaling = (search_space.upper - search_space.lower) / (
+            rescaled_search_space.upper - rescaled_search_space.lower
+        )
+
+        x = tf.clip_by_value(
+            tf.matmul(y, A, transpose_b=True),
+            clip_value_min=-1,
+            clip_value_max=1,
+        )  # project into the new box
+
+        x_rescaled = (
+            x - rescaled_search_space.lower
+        ) * scaling + search_space.lower  # rescale to match the original search space
+
+        return objective(x_rescaled)
```

Comment:
`objective` here requires `d==2`, no? how would then this work for `d==5`? 
(was a bit lazy to run it...)

Reply:
the objective takes the large dimension `D`, not `d`. The above code projects the `d`-dimensional input up to a `D`-dimensional input so it can be used in `objective`